src/main/java/io/luna/game/model/item/shop/Shop.java

```diff
@@ -295,16 +297,159 @@ else if((currentStock - defaultStock) < -10) {
     /**
      * Computes the buy value of {@code item}.
      *
+     * This is based on extensive black-box testing of modern OSRS's shops, and assuming that none of the code for shop prices
+     * has changed since RS2. The principal difference between this function and most private servers is that this one
+     * accounts for the different shop formulas used with items that the store normally stocks
+     * (for example: selling a staff to Zaff, or a security book to a general store).
+     *
      * @param item The item.
-     * @param total If the value should be multiplied by the item's amount.
+     * @param index The index of the item in the shop.
+     * @param amountBought The amount of the item to be purchased.
      * @return The buy value.
      */
-    private int computeBuyValue(Item item, boolean total) {
+    private int computeBuyValue(Item item, int index, int amountBought) {
+        assert amountBought >= 0;
+
         int value = item.getItemDef().getValue();
-        value = value <= 0 ? 1 : value;
-        return total ? value * item.getAmount() : value;
+        int amountStocked = container.computeAmountForIndex(index);
+
+        // TODO: consider maybe moving these somewhere. I'm not sure where. These are just "magic numbers"
+
+        // How much does an item increase in price relative to its base price when understocked by 1 from a specialized store?
+        final double SP_PRICE_CHANGE = 0.02;
+
+        // What is the minimum buy price of an item at a specialized shop?
+        final double SP_MIN_PRICE = 0.1;
+        // What is the maximum buy price of an item at a specialized shop?
+        final double SP_MAX_PRICE = 1.3;
+
+        // How many items need to be understocked to reach max price and how many items need to be overstocked to reach minimum price?
+        final int SP_ITEMS_TO_REACH_MAX_PRICE = (int) ((SP_MAX_PRICE - 1) / SP_PRICE_CHANGE); // 15 by default.
+        final int SP_ITEMS_TO_REACH_MIN_PRICE = (int) ((1 - SP_MIN_PRICE) / SP_PRICE_CHANGE) - 1; // 44 by default. Should be less than the "actual" value of 45 by default.
+
+        // How much does an item increase in price relative to its base price when "understocked" by 1 from a general store?
+        // In this case the "default stock" is treated as 10 even though stock will adjust until there's 0 of the item.
+        final double GEN_PRICE_CHANGE = 0.03;
+
+        // What is the minimum buy price of an item at a general store?
+        final double GEN_MIN_PRICE = 0.3;
+        final int GEN_ITEMS_TO_REACH_MIN_PRICE = (int) ((1 - GEN_MIN_PRICE) / GEN_PRICE_CHANGE); // rounds to 23 by default. Should be less than the "actual" value of 23.33 by default
+
```

Comment:
These local fields should use camelCase naming instead of SNAKE_UPPER_CASE naming. Or keep the naming and make them private class constants